Map List<T> operations to PostgreSQL arrays

austindrenski opened the pull request:
This is a subset of the work from #431 refactored without `IQueryOptimizer` or `NpgsqlExistsToAnyRewritingExpressionVisitor` per #460.

## Compatibility

Operator | PostgreSQL support
-- | --
`=` | 7.4+
`<>` | 7.4+
`<` | 7.4+
`>` | 7.4+
`<=` | 7.4+
`>=` | 7.4+
`@>` | 8.2+
`<@` | 8.2+
`&&` | 8.2+
`\|\|` | 7.4+

Function | PostgreSQL support
-- | --
`array_append(anyarray, anyelement)` | see `\|\|`
`array_cat(anyarray, anyarray)` | see `\|\|`
`array_dims(anyarray)` | 7.4+
`array_fill(anyelement, int[] [, int[]]) ` | 8.4+
`array_length(anyarray, int)` | 8.4+
`array_lower(anyarray, int)` | 7.4+
`array_ndims(anyarray)` | 8.4+
`array_position(anyarray, anyelement [, int])` | 9.5+
`array_positions(anyarray, anyelement)` | 9.5+
`array_prepend(anyelement, anyarray)` | see `\|\|`
`array_remove(anyarray, anyelement)` | 9.3+
`array_replace(anyarray, anyelement, anyelement)` | 9.3+
`array_to_string (anyarray, text)` | 7.4-9.0
`array_to_string(anyarray, text [, text])` | 9.1+
`array_upper(anyarray, int)` | 7.4+
`string_to_array (text, text)` | 7.4-9.0
`string_to_array(text, text [, text])` | 9.1+

### Postponed 
Function | PostgreSQL support
-- | -
`array_agg(expression)` | 8.4+
`cardinality(anyarray)` | 9.4+
`unnest(anyarray)` | 8.4+
`unnest(anyarray, anyarray [, ...])` | 9.4+

## Related
#395
#431 
#460 
#542 
#543 
#544 
#545 
#549 
#551 
#552

austindrenski commented:
Test failures on missing CLR mappings for:
- `uint[]`
- `uint?[]`
- `Guid[]`
- `Guid?[]`

```
SimpleQueryNpgsqlTest.Contains_with_local_int_array_closure

SimpleQueryNpgsqlTest.Contains_with_local_nullable_int_array_closure

GearsOfWarQueryNpgsqlTest.Include_where_list_contains_navigation

GearsOfWarQueryNpgsqlTest.Include_where_list_contains_navigation2

GearsOfWarQueryNpgsqlTest.Contains_with_local_nullable_guid_list_closure

GearsOfWarQueryNpgsqlTest.Navigation_accessed_twice_outside_and_inside_subquery

NullSemanticsQueryNpgsqlTest.Contains_with_local_array_closure_with_multiple_nulls
```

Other test failures:

```
GearsOfWarQueryNpgsqlTest.Contains_on_nullable_array_produces_correct_sql
Error Message:
 Assert.Equal() Failure
Expected: 2
Actual:   1

NullSemanticsQueryNpgsqlTest.Contains_with_local_array_closure_with_null
Error Message:
 Assert.Equal() Failure
Expected: 18
Actual:   9

NullSemanticsQueryNpgsqlTest.Contains_with_local_array_closure_with_multiple_nulls
Error Message:
 Assert.Equal() Failure
Expected: 18
Actual:   9

NullSemanticsQueryNpgsqlTest.Contains_with_local_array_closure_false_with_null
Error Message:
 Assert.Equal() Failure
Expected: 9
Actual:   0

QueryFilterFuncletizationNpgsqlTest.DbContext_list_is_parametrized
Error Message:
 Assert.True() Failure
Expected: True
Actual:   False
```

austindrenski commented:
Translations are implemented for all array functions, with the exception of `unnest(anyarray)` and `unnest(anyarray, anyarray [, ...])`. These two functions are different enough that I think they're best left to a separate PR. (The same goes for `array_agg(expression)`.) 

These could be translated from `Queryable.Select(...)` and `Queryable.Aggregate(...)` when the source is an array/list. This would involve working with `SubQueryExpression` and `SubQueryModel` which are more complicated than the rest of the translations in this PR.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+                return
+                    Coalesce(
+                        new SqlFunctionExpression(
+                            "array_lower",
```
PostgreSQL arrays are 1-based by default unlike .NET arrays which are 0-based by default. So the lower (and upper) bounds have to be adjusted.

Also, why are we coalescing? In what situation will a null be returned?

austindrenski replied:
Ah, I didn't think about needing to adjust incoming (just outgoing)! I'll fix that.

As for the coalesce: 

```
postgres=# SELECT array_lower(ARRAY[]::int[], 1) IS NULL;
 ?column?
----------
 t
(1 row)
```

Whereas the .NET functions are [documented (old docs)](https://msdn.microsoft.com/en-us/library/system.array.getlowerbound(v=vs.110).aspx) to return `-1` when empty. The [new docs](https://docs.microsoft.com/en-us/dotnet/api/system.array.getlowerbound?view=netframework-4.7.2) indicate that a value is always returned, though they no longer list `-1` as the returned value.

roji replied:
Ok, makes sense then (maybe worth confirming the .NET behavior).

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+        [CanBeNull]
+        Expression ListInstanceHandler([NotNull] MethodCallExpression expression)
+        {
+            if (!(expression.Object is Expression instance))
```
Why this check? `MethodCallExpression.Object` already is an expression... If you just want a shorter local variable (`instance`) just assign...

(I think I wrote this elsewhere already)

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+        /// A translated expression or null if no translation is supported.
+        /// </returns>
+        [CanBeNull]
+        static Expression StringInstanceHandler([NotNull] MethodCallExpression expression)
```
As written above, it may be better to move string translations to their own translator.

austindrenski replied:
I'm open to moving this... but it would require some duplication in the `Enumerable` and index handling (see the next comment).

I'm going to hold off moving anything in the next commit, but still open to it if you think the duplication trade-off is acceptable.

roji replied:
Yeah, looking at this again after a while I'd move it to a string translator. It makes more sense to me to have a translator per CLR type (roughly), much easier to discover who does which translations etc.

austindrenski replied:
Okay, sounds good. It fits nicely into the refactored `NpgsqlStringTranslator` (which I don't think we had back in August).

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+        /// An <see cref="IndexExpression"/>.
+        /// </returns>
+        [NotNull]
+        static IndexExpression MakeIndex([NotNull] Expression instance, [NotNull] Expression index)
```
Maybe add to the documentation that this is not used for arrays, only lists.

austindrenski replied:
It's also used for string indexers, but I'll document it all the same.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlCompositeMemberTranslator.cs:
```diff
@@ -50,9 +50,18 @@ public class NpgsqlCompositeMemberTranslator : RelationalCompositeMemberTranslat
             AddTranslators(MemberTranslators);
 
+            // ReSharper disable once DoNotCallOverridableMethodsInConstructor
+            AddTranslators(instanceTranslators);
```
nit: naming isn't ideal: `MemberTranslators` vs `instanceTranslators`. Maybe rename to `versionDependentTranslators`?

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -86,33 +63,15 @@ public class NpgsqlSqlTranslatingExpressionVisitor : SqlTranslatingExpressionVis
-
-            return base.VisitBinary(expression);
-        }
+            => expression.NodeType is ExpressionType.ArrayIndex && IsSafeToVisit(expression, Context)
```
I identify with the desire to use expression-bodied methods, but the `IsSafeToVisit()` method is called only from this method and seems pretty specific to it... So I'd rather this method included its "safety checks" within itself rather than an additional method.

In addition there's multidimensional arrays... Are we supposed to be using `ArrayIndex` for single-dimensional and `ArrayAccess` for multi-dimensional? I understand it may not matter because the expression tree doesn't actually get executed in .NET - it gets translated to SQL - but I think it's still important to keep accurate on these things.

It's also OK to not support multidimensional arrays all in this PR (the current code doesn't).

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -122,101 +81,80 @@ protected override Expression VisitBinary(BinaryExpression expression)
-                fromExpression, _queryModelVisitor.QueryCompilationContext, out _);
+        /// <inheritdoc />
+        protected override Expression VisitUnary(UnaryExpression expression)
+            => Visit(expression.Operand) is Expression operand
```
This seems wrong, isn't `UnaryExpression.Operand` always an expression? That means you're always translating all unary expressions....

austindrenski replied:
`UnaryExpression.Operand` is indeed always an expression, but the result of `Visit(UnaryExpression.Operand)` can be `null` (which `VisitUnary` is not supposed to return). 

Since we need to check the result for `null` before reconstructing/skipping, I would prefer to use pattern matching here. Since pattern matching only makes the variable available in the `true` branch, there's no chance for a `null` result to accidentally be passed back in the `false` branch.

> always translating all unary expressions....

Unless it's an expression we care about (and have written a translator for), I believe that calling `Visit` should either:
1. return `null`, in which case we just move onto the base method; or,
2. return what the base method would have otherwise returned.

To be on the safe-side, I'll re-verify this works as described after the next commit.

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -122,101 +81,80 @@ protected override Expression VisitBinary(BinaryExpression expression)
-            if (lastPropertyType.IsArray && lastPropertyType.GetArrayRank() == 1 && subQueryModel.ResultOperators.Count > 0)
-            {
-                // Translate someArray.Length
-                if (subQueryModel.ResultOperators.First() is CountResultOperator)
```
I'm a bit surprised array indexing works without the special handling that you removed - are we sure the tests are exercising it correctly etc.?

Note that there's a distinction between the translation phase (represented by `NpgsqlSqlTranslatingExpressionVisitor`) and the SQL generation phase. I admit I'm not totally totally clear on the details of this separation, but it seems more correct to have a standard `ArrayLength` node in the expression tree as before. This way, when our SQL generator starts its job, it has an input expression tree that is as close as possible to what a real C# expression tree would look like. We need to review all this and understand it better.

roji commented on src/EFCore.PG/Query/Expressions/Internal/ArrayAnyAllExpression.cs:
```diff
@@ -95,7 +96,9 @@ public class ArrayAnyAllExpression : Expression, IEquatable<ArrayAnyAllExpressio
             => visitor is NpgsqlQuerySqlGenerator npgsqlGenerator
                 ? npgsqlGenerator.VisitArrayAnyAll(this)
-                : base.Accept(visitor);
+                : visitor is NpgsqlSqlTranslatingExpressionVisitor npgsqlVisitor
```
This seems very odd and non-standard... What's the idea?

austindrenski replied:
Looking at this again...I'm not sure.....but it makes things work?

I'll give this a closer look after the next commit. I suspect it's needed due to the translation/generation ~~changes~~ breakages I've introduced.

austindrenski replied:
Ah okay, I remember now:

Per some discussion in #551, this was a way to avoid overriding `VisitExtension(...)` on `NpgsqlSqlTranslatingExpressionVisitor`. 

But looking at it all again, I think we need to be handling this stuff in `VisitExtension(...)` after all. (The [default `SqlTranslatingExpressionVisitor`](https://github.com/aspnet/EntityFrameworkCore/blob/91653bdbe03cd2cbafc6c81c5bad8e1a7d4ef02f/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L1077-L1144) seems to follow such a pattern.)

roji commented on src/EFCore.PG/Query/Expressions/Internal/CustomBinaryExpression.cs:
```diff
@@ -60,7 +61,9 @@ public class CustomBinaryExpression : Expression
             => visitor is NpgsqlQuerySqlGenerator npgsqlGenerator
                 ? npgsqlGenerator.VisitCustomBinary(this)
-                : base.Accept(visitor);
+                : visitor is NpgsqlSqlTranslatingExpressionVisitor npgsqlVisitor
```
This seems very odd and non-standard, and is probably related to this PR deviating from the standard translation -> generation pipeline. `CustomBinaryExpression` is something that's only supposed to exist in the SQL generation phase (which is rather late), and not in the translation phase.

austindrenski replied:
I think you're right, I'll give this another look too.

austindrenski replied:
See my comments above for `ArrayAnyAllExpression`.

roji commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -189,54 +188,52 @@ protected override Expression VisitBinary(BinaryExpression expression)
                 goto default;
             }
 
-            case ExpressionType.ArrayIndex:
```
See comment above, in my opinion this should stay.

roji commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -189,54 +188,52 @@ protected override Expression VisitBinary(BinaryExpression expression)
         /// <inheritdoc />
-        protected override Expression VisitUnary(UnaryExpression expression)
-        {
-            if (expression.NodeType == ExpressionType.ArrayLength)
```
As above, this should stay...

roji commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -189,54 +188,52 @@ protected override Expression VisitBinary(BinaryExpression expression)
+                return
+                    VisitSqlFunction(
+                        new SqlFunctionExpression(
+                            "ascii",
```
C# string subscripting yields a char, which the EF Core provider maps to `text` - so why convert to an int here via `ascii()`?

Regardless, it would be better to not include this kind of change in such a big PR, it's very easy to miss while reviewing...

austindrenski replied:
We discussed this (a while back now) in #450. I haven't found a good solution for identifying when `byte` is really `char`, so I was going to stick with this... But I'm ___very___ open to a better solution, this one just feels so hackish.

austindrenski replied:
I was able to identify this case in `VisitUnary` and apply `ascii` there and out of the indexers.

austindrenski commented on doc/mapping/array.md:
```diff
@@ -7,29 +7,50 @@ PostgreSQL has the unique feature of supporting [*array data types*](https://www
+| `.Where(x => x.SomeArray.Count() == 1)`                          | [`WHERE array_length(x."SomeArray", 1) = 1`](https://www.postgresql.org/docs/current/static/functions-array.html#ARRAY-FUNCTIONS-TABLE)
+| `.Where(x => x.SomeList.Count() == 1)`                           | [`WHERE array_length(x."SomeList", 1) = 1`](https://www.postgresql.org/docs/current/static/functions-array.html#ARRAY-FUNCTIONS-TABLE)
+| `.Where(x => x.SomeArray == x.SomeList)`                         | [`WHERE x."SomeArray" = x."SomeList"`](https://www.postgresql.org/docs/current/static/functions-array.html#ARRAY-FUNCTIONS-TABLE)
+| `.Where(x => x.SomeArray.Equals(x.SomeList))`                    | [`WHERE x."SomeArray" = x."SomeList"`](https://www.postgresql.org/docs/current/static/functions-array.html#ARRAY-FUNCTIONS-TABLE)
```
I'll drop this from the documentation. I've maintained the explicit translation of `SequenceEquals()`, but I think that the `Equals(...)` translation is just included by default. 

I'll need to verify this, but if I have it right (that it's just the EF Core default), would you want to see an explicit non-translation?

roji replied:
I think I would... Arrays are a bit special in this sense: for most types, `Equals()` really is a value comparison, so translating to a database equality check does make sense by default.

It's not terribly important - if you're doing `Equals()` on arrays you're sort of doing things wrong anyway. It's a matter of sticking to the principle of having the behavior as close as possible to the regular .NET client-side behavior.

austindrenski commented on doc/mapping/array.md:
```diff
@@ -7,29 +7,50 @@ PostgreSQL has the unique feature of supporting [*array data types*](https://www
 # Mapping arrays
 
-Simply define a regular .NET array or `List<>` property, and the provider
+Npgsql maps PostgreSQL arrays to generic `T[]` and `List<T>` types:
```
That's an interesting point, I'll keep it in mind!

I have noticed that the `IsGeneric` flag isn't set on arrays, but I assumed this was legacy (szarray type stuff).

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+
+            var type = expression.Arguments[0].Type;
+
+            if (!type.IsArray && type != typeof(string) && !type.IsGenericType && type.GetGenericTypeDefinition() != typeof(List<>))
```
Ah right. I'll take a first pass at fixing this without splitting out strings. If it gets too complicated (or looks to complicated upon review), I'm on board with splitting them out.

austindrenski replied:
In conjunction with the following comment, I've reworked it to the following:

```c#
switch (expression.Method.Name)
{
case nameof(Enumerable.Count) when type == typeof(string):
    return new SqlFunctionExpression("character_length", typeof(int), new[] { expression.Arguments[0] });

case nameof(Enumerable.Count):
    return VersionAtLeast(8, 4)
        ? new SqlFunctionExpression("array_length", typeof(int), new[] { expression.Arguments[0], Expression.Constant(1) })
        : null;
```

roji replied:
Ok, keep going with that approach (without splitting) and we'll see what it looks like.

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+                return VersionAtLeast(9, 4)
+                    ? new SqlFunctionExpression("cardinality", typeof(int), new[] { expression.Arguments[0] })
+                    : VersionAtLeast(8, 4)
+                        ? new SqlFunctionExpression("array_length", typeof(int), new[] { expression.Arguments[0], Expression.Constant(1) })
```
Hmmm... That's a good point. I'll drop `cardinality()` from `Enumerable.Count()`.

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,401 @@
+
+            switch (expression.Method.Name)
+            {
+            case nameof(Array.IndexOf) when VersionAtLeast(9, 5) && expression.Method.GetParameters().Length <= 3:
```
From what I can see, I think that `Array.IndexOf` is only intended to target one-dimensional arrays:

```xml
/// <exception cref="T:System.RankException"><paramref name="array">array</paramref> is multidimensional.</exception>
```

If someone passes a multidimensional array (and the compiler allows it), is it our responsibility to leave it to throw on the client? Or should we let it pass and allow PostgreSQL to throw?

roji replied:
Yeah, it's true that this is a question of where to throw... `Array.IndexOf()` only accepts multidimensional parameters (and then throws) because they're the same type statically (`Array`). I'd rather we didn't translate and allowed the client to throw.

roji commented:
Answered some of your comments, let me know when you want another review. And thanks for all the valuable work!

austindrenski commented:
_moving some discussion down here for clarity:_

@roji [wrote](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/541#discussion_r207113751):
> Note that there's a distinction between the translation phase (represented by `NpgsqlSqlTranslatingExpressionVisitor`) and the SQL generation phase. I admit I'm not totally totally clear on the details of this separation, but it seems more correct to have a standard `ArrayLength` node in the expression tree as before. This way, when our SQL generator starts its job, it has an input expression tree that is as close as possible to what a real C# expression tree would look like. We need to review all this and understand it better.

@roji [wrote](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/541#discussion_r207114476):
> This seems very odd and non-standard, and is probably related to this PR deviating from the standard translation -> generation pipeline. `CustomBinaryExpression` is something that's only supposed to exist in the SQL generation phase (which is rather late), and not in the translation phase.

So, first, I want to clarify that __I do agree__ that 3a92b465c68a06f0e6173c960cd52105f411176e introduced some blurriness between translation and generation phases, and I'm in the process of cleaning some of that up.

But, from looking at examples in the EF Core repo, I don't _think_ that our custom expressions fit into the translation/generation distinction in _quite_ the way you describe...

For example, [`SqlTranslatingExpressionVisitor`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L636) calls into the `ICompositeMethodCallTranslator` which is definitely allowed to return SQL specific expressions, as seen in various EF Core translators:  
- `SqlServerDateAddTranslator` returns [`SqlFunctionExpression`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerDateAddTranslator.cs#L57) and [`SqlFragmentExpression`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerDateAddTranslator.cs#L62)
- `SqlServerMathTranslator` returns [`ExplicitCastExpression`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerMathTranslator.cs#L108).

Indeed, there are a number of translations we're handling that would be very difficult to achieve without injecting custom expressions prior to the SQL generation phase. 

Basically, I think that we need to allow custom expressions (including PostgreSQL specific expressions) to be injected as soon as `NpgsqlSqlTranslatingExpressionVisitor` starts processing. This should of course be done within reason. (For example, I mistakenly used a `SqlFunctionExpression` for `COALESCE` when I could have just used `Expression.Coalesce`.)

austindrenski commented:
@roji I've split out some of the unrelated changes and rebased. The last review commented on changes to the query pipeline, but I think most of those should be reverted here. 

This PR still doesn't address multidimensional arrays, but given the current size of it, I'd prefer to punt multidimensional support (as well as `unnest` and `array_agg`) to the next release (i.e. `v2.3` or `v3.0`).

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,408 @@
+            switch (expression.Method.Name)
+            {
+            case nameof(NpgsqlArrayExtensions.Contains):
+                return VersionAtLeast(8, 2)
```
(Sorry if we've already discussed this and I forgot)

I'm not sure we should throw here. The extension methods (e.g. `NpgsqlArrayExtensions.Contains`) already throw a ClientEvaluationNotSupportedException, so you can simply translate if the version is correct, and return null otherwise. This will trigger client evaluation which will throw.

In other words, if a database version doesn't support an operation, I think it's better to not translate and let client evaluation take its course.

This will also make code here more terse, as you can move the `VersionAtLeast()` check into the case and remove the conditional expression.

austindrenski replied:
Oh, that's a great point. I'll make the changes throughout.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,408 @@
+            case nameof(NpgsqlArrayExtensions.StringToArray):
+                return VersionAtLeast(9, 1)
+                    ? new SqlFunctionExpression("string_to_array", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(3))
+                    : VersionAtLeast(7, 4)
```
Uhh, as above, supporting pre-9.1 PostgreSQL seems a bit too much... Every bit of code has its price in term of maintenance/complexity etc., at the very least we should wait for users to request support for such old versions rather than eagerly implementing it...

austindrenski replied:
Here I was thinking about PostgreSQL-like systems (e.g. Redshift). That said, this is an Npgsql-specific extension method, so anyone seeking it out should know better... I'll rework this to omit 7.4 checks.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs:
```diff
@@ -0,0 +1,408 @@
+        /// An <see cref="T:ExpressionType.ArrayIndex"/> or <see cref="T:ExpressionType.Index"/>.
+        /// </returns>
+        [NotNull]
+        static Expression MakeIndex([NotNull] Expression instance, [NotNull] Expression index)
```
I'd split to two methods, one for array and another for list (string isn't needed if we move it out to another translator).

austindrenski replied:
I'm just going to omit this altogether and inline the code for arrays/lists.

austindrenski replied:
Scratch that, still need the methods, but will split.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlCompositeMemberTranslator.cs:
```diff
@@ -26,8 +26,16 @@ public class NpgsqlCompositeMemberTranslator : RelationalCompositeMemberTranslat
+            // ReSharper disable once DoNotCallOverridableMethodsInConstructor
             AddTranslators(MemberTranslators);
+
+            // ReSharper disable once DoNotCallOverridableMethodsInConstructor
```
nit: remove space and Resharper comment to put `AddTranslators()` lines together. Can use a Resharper region disable for the warning (begin/end).

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -20,39 +18,15 @@ namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionVisitors
         /// </summary>
-        [NotNull] readonly RelationalQueryModelVisitor _queryModelVisitor;
+        [NotNull]
+        RelationalQueryCompilationContext Context => _queryModelVisitor.QueryCompilationContext;
```
Probably a good idea to maintain the full name (`QueryCompilationContext`) since `Context` is quite ambiguous within EF Core (e.g. `DbContext`).

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -64,29 +38,24 @@ public class NpgsqlSqlTranslatingExpressionVisitor : SqlTranslatingExpressionVis
         protected override Expression VisitBinary(BinaryExpression expression)
         {
-            if (expression.NodeType == ExpressionType.ArrayIndex)
+            var left = Visit(expression.Left);
```
The point of the previous code was to immediately delegate to `base.VisitBinary()` in all cases, *unless* when the expression's node type is `ArrayIndex`. In the new code you call `Visit()` on Left and Right, and then it gets called again if the expression isn't `ArrayIndex` or `Index`. This could be bad for performance and may potentially have other side-effects - when we're changing the query pipeline I'd rather be as little invasive as we can.

So I'd start with a switch on the NodeType, which immediately returns `base.VisitBinary()` for anything but what we handle here.

austindrenski replied:
I've reworked this into a switch to cut down on the possibility of wasted visits. How does it look now?

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -100,105 +69,117 @@ protected override Expression VisitBinary(BinaryExpression expression)
+            if (expression.NodeType == ExpressionType.Convert &&
+                Visit(expression.Operand) is IndexExpression index &&
+                index.Object.Type == typeof(string))
+                return new SqlFunctionExpression("ascii", typeof(int), new[] { index });
```
Isn't the `substr` invocation missing? `ascii` only returns the code of the first character of the string...

austindrenski replied:
Nope, the substring is generated via the `IndexExpression`, so here we just have to make sure it's wrapped with `ascii` to handle the odd `character <-> integer` conversion. It looks a bit convoluted, but there are tests that verify the behavior.

roji commented on src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -100,105 +69,117 @@ protected override Expression VisitBinary(BinaryExpression expression)
+        {
+            switch (expression)
+            {
+            case ArrayAnyAllExpression e:
```
I'm confused... `NpgsqlArrayFragmentTranslator` now constructs and returns `ArrayAnyAllExpression`s, are they expected to find their way back here in the SqlTranslatingExpressionVisitor? Is this because we assume the fragment translators execute before this method? Either way, the whole point of having the FragmentTranslator is to cleanly encapsulate all the array logic in one place, but this seems to spread it back around...

The situation seems to be similar with the other custom expression we have, but I may be misunderstanding completely.

austindrenski replied:
I based this on the [implementation in SqlTranslatingExpressionVisitor](https://github.com/aspnet/EntityFrameworkCore/blob/8e7402701ea1416b7fe55e1863ae65b3147c02c7/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L1091-L1158) which does similar work on `NodeType.Extension` expressions defined upstream.

The fragment translators [always run first](https://github.com/aspnet/EntityFrameworkCore/blob/8e7402701ea1416b7fe55e1863ae65b3147c02c7/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L103-L105), so the idea is to package up a bunch of complicated nodes into one custom expression that we can easily identify and handle later on in the normal translator workflow.

It does spread around _some_ logic, but this is still advantageous for the `NpgsqlSqlTranslatingExpressionVisitor`, because we're just performing normal expression visitor operations on `ArrayAnyAllExpression` (e.g. visit members, return original or modified).

_Note we do have one remaining piece of complicated `ArrayAnyAllExpression` logic outside the fragment translator, which involves translating a `ContainsResultOperator`. I want to refactor that into a fragment translator too, but I'd like to deal with that in a later PR._

roji commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -202,31 +202,52 @@ protected override Expression VisitUnary(UnaryExpression expression)
+                        new[] { expression.Object, GenerateOneBasedIndexExpression(expression.Arguments[0]), Expression.Constant(1) }));
+            }
+
+            Visit(expression.Object);
```
I don't think we're sure that we're dealing with an array at this point... I think a check is missing (immediately return null otherwise)

austindrenski commented:
@roji Thanks for the review. I'm (slowly) working my way through it, and hope to have it addressed within the next day or two.

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMemberTranslator.cs:
```diff
@@ -0,0 +1,140 @@
+        {
+            var instance = expression.Expression;
+
+            if (instance is null || !instance.Type.IsArray)
```
Yes, I've updated accordingly. I'm working on stripping the (haphazard) multidimensional support now.

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMemberTranslator.cs:
```diff
@@ -0,0 +1,140 @@
+            case nameof(Array.Length) when VersionAtLeast(9, 4):
+                return new SqlFunctionExpression("cardinality", typeof(int), new[] { instance });
+
+            case nameof(Array.Length) when VersionAtLeast(8, 4) && instance.Type.GetArrayRank() == 1:
```
Actually, after some more thought: I can't really define what value these checks are adding for unsupported versions, so I'm going to drop the `VersionAtLeast` checks for things prior to `9.4`. 

This means that, going forward, we should be able to say that translators will only run if the version of PostgreSQL is currently supported and have this "minimum version for translations" increment with the support cycle.

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMemberTranslator.cs:
```diff
@@ -0,0 +1,140 @@
+        /// A translated expression or null if no translation is supported.
+        /// </returns>
+        [CanBeNull]
+        Expression ArrayInstanceHandler([NotNull] MemberExpression expression)
```
I find it a bit surprising too, but from everything I'm seeing, it is indeed a `MemberExpression`.

When debugging `ArrayQueryTest`. I can see that the array property is actually wrapped up as a `ColumnExpression`, so perhaps that has something to do with it:

![image](https://user-images.githubusercontent.com/21338699/49850513-c28c2700-fdab-11e8-9d91-116ccf971e1c.png)

roji commented:
@austindrenski I just re-noticed this... It's really unfortunately we didn't get this merged before the new query pipeline work. Do you have time to rebase this to the latest version and do the porting?

darbio commented:
Any movement on getting this into the library? It would be good to not have to worry about arrays T[] and just use List<T> in the code base I'm working on.

roji commented:
Closing as there has been no comment from @austindrenski for a while, and this PR would probably need to be redone anyway after the query pipeline redo of 3.0. The tracking issue here is #395.
